c7n/resources/cw.py

```diff
@@ -114,6 +119,135 @@ def get_dimensions(self, resource):
+        arn_map = {}
+        r_map = {}
+        child_resources = self.get_related(resources)
+        for r in resources:
+            r_map[r['Name']] = r
+        for k, v in child_resources.items():
+            for r in v:
+                arn_map.setdefault(k, {}).setdefault('Arns', []).append(r['Arn'])
+                arn_map[k]['data'] = r_map[v[0]['c7n:parent-id']]
+        return arn_map
+
+    def process(self, resources, event=None):
+        # Due to lazy loading of resources, we need to explicilty load the following
+        # potential targets for a event rule target:
+        load_resources(
+            [
+                "aws.sqs",
+                "aws.lambda",
+                "aws.ecs-cluster",
+                "aws.ecs-task",
+                "aws.kinesis",
+                "aws.sns",
+                "aws.ssm-parameter",
+                "aws.batch-compute",
+                "aws.codepipeline",
+            ]
+        )
+        arn_resolver = ArnResolver(self.manager)
+        arn_map = self._get_arns(resources)
+        results = []
+        for k, v in arn_map.items():
+            resolved = arn_resolver.resolve(v['Arns'])
+            if not all(resolved.values()):
+                for i, j in resolved.items():
+                    if not j:
+                        v['data'].setdefault('c7n:InvalidTargets', []).append(i)
+                results.append(v['data'])
+        return results
+
+
+@EventRule.action_registry.register('delete')
+class EventRuleDelete(BaseAction):
+    """
+    Delete an event rule, force target removal with the `force` option
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+            - name: force-delete-rules
+              resource: aws.event-rule
+              filters:
+                - Name: my-event-rule
+              actions:
+                - type: delete
+                  force: true
+    """
+
+    schema = type_schema('delete', force={'type': 'boolean'})
+    permissions = ('events:DeleteRule', 'events:RemoveTargets', 'events:ListTargetsByRule',)
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('events')
+        children = {}
+        target_error_msg = "Rule can't be deleted since it has targets."
+        for r in resources:
+            try:
+                client.delete_rule(Name=r['Name'])
+            except botocore.exceptions.ClientError as e:
+                if self.data.get('force') and e.response['Error']['Message'] == target_error_msg:
+                    child_manager = EventRuleTarget(self.manager.ctx, {})
+                    if not children:
+                        children = EventRuleTargetFilter({}, child_manager).get_related(resources)
+                    targets = list(set([t['Id'] for t in children.get(r['Name'])]))
+                    client.remove_targets(Rule=r['Name'], Ids=targets)
+                    client.delete_rule(Name=r['Name'])
```

Comment:
both the delete api calls should be wrapped with utils retry and pass ignore_err_code for the not found value if the resource is already deleted

Reply:
Looks like when deleting a non-existent resource it actually returns a 200:

```
>>> import boto3
>>> client = boto3.client('events')
>>> client.delete_rule(Name='foobarbaz')
{'ResponseMetadata': {'RequestId': '0a2b0651-b300-45cd-a4e6-7b89525d96f0', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': '0a2b0651-b300-45cd-a4e6-7b89525d96f0', 'content-type': 'application/x-amz-json-1.1', 'content-length': '0', 'date': 'Wed, 03 Mar 2021 15:12:20 GMT'}, 'RetryAttempts': 0}}
```

Reply:
same with removing targets

```
>>> client.remove_targets(Rule='test', Ids=['foo'])
{'FailedEntryCount': 0, 'FailedEntries': [], 'ResponseMetadata': {'RequestId': 'f0a3b614-6516-4079-b535-0175221a389e', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': 'f0a3b614-6516-4079-b535-0175221a389e', 'content-type': 'application/x-amz-json-1.1', 'content-length': '41', 'date': 'Wed, 03 Mar 2021 18:05:54 GMT'}, 'RetryAttempts': 0}}
```

Reply:
thanks for verifying